src/main/java/com/microsoft/appcenter/espresso/ReportHelper.java

```diff
@@ -2,48 +2,77 @@
 
 import com.microsoft.appcenter.event.EventReporter;
 import com.microsoft.appcenter.event.EventType;
+
+import net.jcip.annotations.GuardedBy;
+
 import org.junit.rules.TestWatcher;
 import org.junit.runner.Description;
 
+import java.util.IdentityHashMap;
+
 public class ReportHelper extends TestWatcher {
+    private final static Object currentReportHelperLock = new Object();
+    @GuardedBy("currentReportHelperLock")
+    private final static IdentityHashMap<ReportHelper, Void> oldReportHelpers = new IdentityHashMap<>();
+    @GuardedBy("currentReportHelperLock")
+    private static ReportHelper currentReportHelper = null;
+
     private final EventReporter eventReporter;
 
     ReportHelper(EventReporter eventReporter) {
         super();
         this.eventReporter = eventReporter;
     }
 
+    private void ensureOnlyOneReporter() {
+       synchronized (currentReportHelperLock) {
+           if(this != currentReportHelper) {
+               if(oldReportHelpers.containsKey(this)) {
+                   throw new IllegalStateException("More that on ReportHelper is active, this will cause report generation to fail. One common cause for this @Rules in test class and in its super class");
+               } else {
+                   oldReportHelpers.put(currentReportHelper, null);
+                   currentReportHelper = this;
+               }
+           }
+       }
+    }
+
```

Comment:
While I believe ensureOnlyOneReporter() looks like it does as intended I think it is overly complicated. I don't think we need to hold references to all old reporters, and I don't think we need the jcip dependency, and I don't think we need manual locking - and why use IdentityHashMap instead of a plain old Set? :-) AFAICS just using a synchronized static method and and a sequence number should suffice... something ala:

```java
private int reporterId = -1;
private static int reporterIdCounter = 0;

private static synchronized void ensureIsCurrentReporter(ReportHelper self) {
  if (self.reporterId == -1) {
    self.reporterId = ++reporterIdCounter;
  }
  
  if (self.reportId != reportedIdCounter) {
    throw new IllegalStateException("More than one ReportHelper is active, this will " + 
      "cause report generation to fail. One common cause for this @Rules in test class " +
      "and in its super class");
  }
}
```